datafusion/expr/src/window_frame.rs

```diff
@@ -334,51 +335,69 @@ impl WindowFrameBound {
     }
 }
 
-impl TryFrom<ast::WindowFrameBound> for WindowFrameBound {
-    type Error = DataFusionError;
-
-    fn try_from(value: ast::WindowFrameBound) -> Result<Self> {
+impl WindowFrameBound {
+    fn try_parse(
+        value: ast::WindowFrameBound,
+        units: &ast::WindowFrameUnits,
+    ) -> Result<Self> {
         Ok(match value {
             ast::WindowFrameBound::Preceding(Some(v)) => {
-                Self::Preceding(convert_frame_bound_to_scalar_value(*v)?)
+                Self::Preceding(convert_frame_bound_to_scalar_value(*v, units)?)
             }
             ast::WindowFrameBound::Preceding(None) => Self::Preceding(ScalarValue::Null),
             ast::WindowFrameBound::Following(Some(v)) => {
-                Self::Following(convert_frame_bound_to_scalar_value(*v)?)
+                Self::Following(convert_frame_bound_to_scalar_value(*v, units)?)
             }
             ast::WindowFrameBound::Following(None) => Self::Following(ScalarValue::Null),
             ast::WindowFrameBound::CurrentRow => Self::CurrentRow,
         })
     }
 }
 
-pub fn convert_frame_bound_to_scalar_value(v: ast::Expr) -> Result<ScalarValue> {
-    Ok(ScalarValue::Utf8(Some(match v {
-        ast::Expr::Value(ast::Value::Number(value, false))
-        | ast::Expr::Value(ast::Value::SingleQuotedString(value)) => value,
-        ast::Expr::Interval(ast::Interval {
-            value,
-            leading_field,
-            ..
-        }) => {
-            let result = match *value {
-                ast::Expr::Value(ast::Value::SingleQuotedString(item)) => item,
-                e => {
-                    return sql_err!(ParserError(format!(
-                        "INTERVAL expression cannot be {e:?}"
-                    )));
+pub fn convert_frame_bound_to_scalar_value(
+    v: ast::Expr,
+    units: &ast::WindowFrameUnits,
+) -> Result<ScalarValue> {
+    match units {
+        // For ROWS and GROUPS we are sure that the ScalarValue must be a non-negative integer ...
+        ast::WindowFrameUnits::Rows | ast::WindowFrameUnits::Groups => match v {
+            ast::Expr::Value(ast::Value::Number(value, false))
+            | ast::Expr::Value(ast::Value::SingleQuotedString(value)) => {
```

Comment:
I tried to add some sqllogictests case for `SinglQuotedString` but I found it doesn't work in end-to-end SQL test.
```
External error: statement failed: DataFusion error: Error during planning: Invalid window frame: frame offsets for ROWS / GROUPS must be non negative integers
[SQL] select row_number() over (rows between '1' preceding and current row) from (select 1 a) x
at test_files/window.slt:2381
```

I think this worked in the previous version but it isn't covered in the tests. 🤔  
```
DataFusion CLI v42.0.0
> select row_number() over (rows between '1' preceding and current row) from (select 1 a) x;
+-------------------------------------------------------+
| row_number() ROWS BETWEEN 1 PRECEDING AND CURRENT ROW |
+-------------------------------------------------------+
| 1                                                     |
+-------------------------------------------------------+
1 row(s) fetched. 
Elapsed 0.001 seconds.
```

I think we should add some `SingleQuotedString` case in the sqllogictests.

Reply:
Great catch!

The `rows between '1' preceding and current row` maps to the following AST expression.
```
Interval(
    Interval {
        value: Value(
            SingleQuotedString(
                "1",
            ),
        ),
        leading_field: None,
        leading_precision: None,
        last_field: None,
        fractional_seconds_precision: None,
    },
)
```

It seems like for `Groups` we need to also handle the `ast::Expr::Interval(...)` pattern to preserve existing behaviour.

Reply:
Curiously, do you know what SQL (or substrait?) will generate this case, Rows with SingleQuotedString? I'm wondering if we should keep this matching rule or remove it.

Reply:
Curiously I'm not sure it will show up. It will either be Interval or Value(Number). So I guess we can remove it.

Reply:
The relevant parsing code is here: https://github.com/apache/datafusion-sqlparser-rs/blob/8e0d26abb3c4dda2fa091dfd9611be1dbdc14755/src/parser/mod.rs#L1483-L1490

Reply:
This means a single quoted string will always be parsed as an interval 🤷